projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/intrinsics/rust/LLVMStart.java

```diff
@@ -0,0 +1,202 @@
+import com.oracle.truffle.llvm.runtime.LLVMFunctionDescriptor;
+import com.oracle.truffle.llvm.runtime.datalayout.DataLayout;
+import com.oracle.truffle.llvm.runtime.global.LLVMGlobal;
+import com.oracle.truffle.llvm.runtime.memory.LLVMMemory;
+import com.oracle.truffle.llvm.runtime.memory.LLVMStack.StackPointer;
+import com.oracle.truffle.llvm.runtime.nodes.api.LLVMExpressionNode;
+import com.oracle.truffle.llvm.runtime.nodes.api.LLVMNode;
+import com.oracle.truffle.llvm.runtime.nodes.api.LLVMToNativeNode;
+import com.oracle.truffle.llvm.runtime.pointer.LLVMNativePointer;
+import com.oracle.truffle.llvm.runtime.pointer.LLVMPointer;
+import com.oracle.truffle.llvm.runtime.types.FunctionType;
+import com.oracle.truffle.llvm.runtime.types.StructureType;
+import com.oracle.truffle.llvm.runtime.types.Type;
+
+public abstract class LLVMStart extends LLVMIntrinsic {
+
+    protected LLVMClosureDispatchNode getClosureDispatchNode() {
+        CompilerAsserts.neverPartOfCompilation();
+        return LLVMClosureDispatchNodeGen.create();
+    }
+
+    @NodeChildren({@NodeChild(type = LLVMExpressionNode.class), @NodeChild(type = LLVMExpressionNode.class), @NodeChild(type = LLVMExpressionNode.class), @NodeChild(type = LLVMExpressionNode.class)})
+    public abstract static class LLVMLangStart extends LLVMStart {
+
+        @Specialization
+        @SuppressWarnings("unused")
+        protected long doOp(StackPointer stackPointer, LLVMNativePointer main, long argc, LLVMPointer argv,
+                        @Cached("getClosureDispatchNode()") LLVMClosureDispatchNode dispatchNode) {
+            dispatchNode.executeDispatch(main, new Object[]{stackPointer});
+            return 0;
+        }
+
+        @Specialization
+        @SuppressWarnings("unused")
+        protected long doOp(StackPointer stackPointer, LLVMFunctionDescriptor main, long argc, LLVMPointer argv,
+                        @Cached("getClosureDispatchNode()") LLVMClosureDispatchNode dispatchNode) {
+            dispatchNode.executeDispatch(main, new Object[]{stackPointer});
+            return 0;
+        }
+
+    }
+
+    @NodeChildren({@NodeChild(type = LLVMExpressionNode.class), @NodeChild(type = LLVMExpressionNode.class), @NodeChild(type = LLVMExpressionNode.class), @NodeChild(type = LLVMExpressionNode.class),
+                    @NodeChild(type = LLVMExpressionNode.class)})
+    public abstract static class LLVMLangStartInternal extends LLVMStart {
+
+        protected LangStartVtableType createLangStartVtable(Type vtableType) {
+            DataLayout dataSpecConverter = getContextReference().get().getDataSpecConverter();
+            return LangStartVtableType.create(dataSpecConverter, vtableType);
+        }
+
+        @Specialization
+        @SuppressWarnings("unused")
+        protected long doOp(StackPointer stackPointer, LLVMNativePointer mainPointer, LLVMGlobal vtable, long argc, LLVMPointer argv,
```

Comment:
This code seems to rely on the fact that it is executed only once (you are caching values that are derived from the method arguments, without using any guards).
- if it is guaranteed that this code is only executed once, the caching will only be an overhead and won't have any benefit as the code is never compiled
- if it is not guaranteed that this code is only executed once, you will need guards

Reply:
It indeed relies on the fact that it is executed only once. `lang_start_internal` is called exactly once at program start. I removed most of the caching in my latest commit. I am unsure whether it makes sense to cache `LLVMMemory`. Its cached value is not derived from method arguments, but I guess I should still get rid of it for the reason that it is mere overhead.

Reply:
Thanks for cleaning that up. Yes, you could also get rid of caching `LLVMMemory`. Besides that, the change looks good to me.